plenum/server/consensus/view_change_service.py

```diff
@@ -212,14 +244,167 @@ def _validate(self, msg: Union[ViewChange, ViewChangeAck, NewView], frm: str) ->
+        self._data.checkpoints.append(cp)
+
+        # Update batches
+        # TODO: Actually we'll need to retrieve preprepares by ID from somewhere
+        self._data.preprepared = batches
+
+        # We finished a view change!
         self._data.waiting_for_new_view = False
+
+    def _clear_old_batches(self, batches: Dict[int, Any]):
+        for pp_seq_no in list(batches.keys()):
+            if pp_seq_no <= self._data.stable_checkpoint:
+                del batches[pp_seq_no]
+
+    @staticmethod
+    def _batch_id(batch: PrePrepare):
+        return BatchID(batch.viewNo, batch.ppSeqNo, batch.digest)
+
+    def _calc_checkpoint(self, vcs: List[ViewChange]) -> Optional[Checkpoint]:
+        checkpoints = []
+        for cur_vc in vcs:
+            for cur_cp in cur_vc.checkpoints:
+                # Don't add checkpoint to pretending ones if it is already there
+                if cur_cp in checkpoints:
+                    continue
+
+                # Don't add checkpoint to pretending ones if too many nodes already stabilized it
+                # TODO: Should we take into account view_no as well?
+                stable_checkpoint_not_higher = [vc for vc in vcs if cur_cp.seqNoEnd >= vc.stableCheckpoint]
+                if not self._data.quorums.strong.is_reached(len(stable_checkpoint_not_higher)):
+                    continue
+
+                # Don't add checkpoint to pretending ones if not enough nodes have it
+                have_checkpoint = [vc for vc in vcs if cur_cp in vc.checkpoints]
+                if not self._data.quorums.weak.is_reached(len(have_checkpoint)):
+                    continue
+
+                # All checks passed, this is a valid candidate checkpoint
+                checkpoints.append(cur_cp)
+
+        highest_cp = None
+        for cp in checkpoints:
+            # TODO: Should we take into account view_no as well?
+            if highest_cp is None or cp.seqNoEnd > highest_cp.seqNoEnd:
+                highest_cp = cp
+
+        return highest_cp
+
+    def _calc_batches(self, cp: Checkpoint, vcs: List[ViewChange]) -> List[BatchID]:
+        # TODO: Optimize this
+        batches = set()
+        for vc in vcs:
+            for _bid in vc.prepared:
+                bid = BatchID(*_bid)
+                if bid in batches:
+                    continue
+                if self._is_batch_prepared(bid, cp, vcs):
+                    batches.add(bid)
+
+            for _bid in vc.preprepared:
+                bid = BatchID(*_bid)
+                if bid in batches:
+                    continue
+                if self._is_batch_preprepared(bid, cp, vcs):
+                    batches.add(bid)
+
+        return sorted(batches)
+
+    def _is_batch_prepared(self, bid: BatchID, cp: Checkpoint, vcs: List[ViewChange]) -> bool:
+        if not self._is_inside_watermarks(bid, cp):
+            return False
+
+        def check(vc: ViewChange):
+            if bid.pp_seq_no <= vc.stableCheckpoint:
```

Comment:
This is the condition `m'.h < n`, so that it must be `vc.stableCheckpoint < bid.pp_seq_no`